### DISCUSSION TITLE
[Chat] Powershell script in Custom Chat Application?

### DISCUSSION THREAD

**Discussion by KaXaSA** [streamlink/streamlink-twitch-gui#903] [Q&A]
I want to run a PowerShell script to do some stuff before/after running the chat client.

This works:
```
Application executable: 
cmd.exe

Application parameters: 
/c "powershell -ExecutionPolicy ByPass C:\Folder\Script.ps1 -channel '{channel}'"
```

But I was wondering why if I try to use PowerShell directly, it doesn't work:

```
Application executable: 
powershell.exe

Application parameters: 
-File "C:\Folder\script.ps1" -ExecutionPolicy ByPass -channel "{channel}"
```

Thanks!

---

Edit: 

Using, this also works:
```
Application executable: 
cmd.exe

Application parameters: 
/c powershell -ExecutionPolicy Bypass -File "C:\Folder\Chat Launcher.ps1" -channel "{channel}"

```
This is better because when the file/folders path have any spaces you need to place " around it. In the initial example, that was not possible. (at least everything I've tried failed to work)

---

To make it work with **PowerShell Preview** I had to do this:

```
Application parameters: 
/c ( "C:\Program Files\PowerShell\7-preview\pwsh.exe" -ExecutionPolicy Bypass -File "C:\Folder\Script Test.ps1" -channel "{channel}" )
```

Notice that there is a space after: '(' and before: ')'
That's intentional, if you remove them the command doesn't work at all, super weird.
I'm 100% sure it's related to the space in 'Program Files' because if I use the short version of the path it works fine:
`/c C:\PROGRA~1\PowerShell\7-preview\pwsh.exe -ExecutionPolicy Bypass "C:\Folder\Script Test.ps1" -channel "{channel}"`

(the paths are the same)

Other useful arguments:

-WindowStyle Hidden 
hide the console window (it still shows up for a second, but it's good enough)

-NoExit
Doesn't close the console, useful to see errors that close the console before you can see anything

**Comment by bastimeyer** [comment]
> it doesn't work

That's not very descriptive. You are also using two different launch parameters when executing `powershell` / `powershell.exe`, including different script paths. You're not even setting the same executable, so theoretically, `cmd.exe` could resolve a different powershell executable, which will very likely not be the case, but still...

As you can see [here](https://github.com/streamlink/streamlink-twitch-gui/blob/v2.1.0/src/app/services/chat/launch.js), when Streamlink Twitch GUI launches the chat application, it completely ignores any IO streams and spawns the child process in [detached mode](https://nodejs.org/api/child_process.html#optionsdetached) and it [unreferences](https://nodejs.org/api/child_process.html#subprocessunref) it. And that's already it.

You can see the resolved launch arguments when starting Streamlink Twitch GUI with the [`--loglevel=debug`](https://github.com/streamlink/streamlink-twitch-gui/wiki/Parameters#--loglevel--l) parameter. The `args` get tokenized ([1](https://github.com/streamlink/streamlink-twitch-gui/blob/v2.1.0/src/app/utils/parameters/ParameterCustom.js#L37-L131), [2](https://github.com/streamlink/streamlink-twitch-gui/blob/v2.1.0/src/test/tests/utils/parameters/ParameterCustom.js)) and are passed to NodeJS's `spawn` function accordingly.

I'm not a Windows user and don't know much powershell, so I can't help you with the script. Launching it manually with the parameters resolved by the Twitch GUI and adding debug messages should be the first step that you should do in order to figure this out.

**Reply by KaXaSA** [reply]
Sorry, by 'doesn't work' I mean that the PowerShell console doesn't open at all.
The path, I forgot to change when posting here, the path to the script was the same on both scenarios when I was testing it, that is not the problem.

I used powershell.exe as an example (both with or without '.exe' works in the cmd.exe test), but I actually tried a few things, including the full path, so:
```
powershell
powershell.exe
C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe
C:\Program Files\PowerShell\7-preview\pwsh.exe
```

You can add just the PowerShell full path, **without any arguments**, and PowerShell will still not open. (not sure if intended)

I will check the links tomorrow and try --loglevel=debug to see if I can spot anything, thanks.

**Reply by KaXaSA** [reply]
Ok, so using --loglevel=debug

Powershell - Some Arguments between " (The console doesn't open)

```
Resolving chat provider
{
    "provider": "custom",
    "providerUserData": {
        "exec": "C:\\WINDOWS\\system32\\WindowsPowerShell\\v1.0\\powershell.exe",
        "args": "-File \"C:\\Folder\\Script.ps1\" -ExecutionPolicy ByPass -channel \"{channel}\"",
        "url": "default",
        "type": "settings-chat-provider-custom"
    }
}
```
---

Powershell - Removed all the " from the arguments (The console doesn't open)
```
Resolving chat provider
{
    "provider": "custom",
    "providerUserData": {
        "exec": "C:\\WINDOWS\\system32\\WindowsPowerShell\\v1.0\\powershell.exe",
        "args": "-File C:\\Folder\\Script.ps1 -ExecutionPolicy ByPass -channel {channel}",
        "url": "default",
        "type": "settings-chat-provider-custom"
    }
}
```
---

CMD - Without arguments (The console doesn't open)
```
Resolving chat provider
{
    "provider": "custom",
    "providerUserData": {
        "exec": "cmd.exe",
        "args": "",
        "url": "default",
        "type": "settings-chat-provider-custom"
    }
}
```
---

CMD - With arguments (The console opens fine, and the script works as intended)
By opens fine I mean that PowerShell console opens up, and the script runs fine, I can see the channel, it does some stuff then launch the chat client.
```
Resolving chat provider
{
    "provider": "custom",
    "providerUserData": {
        "exec": "cmd.exe",
        "args": "/c \"powershell -ExecutionPolicy ByPass C:\\Folder\\Script.ps1 -channel '{channel}'\"",
        "url": "default",
        "type": "settings-chat-provider-custom"
    }
}
```

---

So, it only works if you use **cmd.exe** (C:\WINDOWS\system32\cmd.exe) with **/c** (Carries out the command specified by string and then terminates), everything else I've tried just doesn't open at all.

---

Not sure if 100% related, since it's a closed issue from 2017, but I found someone who had a very similar problem: [#385](https://github.com/streamlink/streamlink-twitch-gui/issues/385)
> After messing around a little turned out PS in GUI works correctly **only** if invoked with `cmd.exe /c "powershell .\myscript.ps1"`.

**Reply by bastimeyer** [reply]
The problem comes from the `detached: true` option when spawning child processes in NodeJS:
https://nodejs.org/api/child_process.html#optionsdetached
> On Windows, setting options.detached to true makes it possible for the child process to continue running after the parent exits. The child will have its own console window.

This is necessary for the chat application to continue running when the Twitch GUI gets closed. The same applies to the streamlink/python process for being able to continue watching streams after closing the Twitch GUI.

Recently, there has been a similar issue in regards to the executables generated by python-distlib 0.3.4, where those executables could not be run from NodeJS:
https://github.com/takluyver/pynsist/issues/243

----

See this for example:

**foo.js**
```js
const { spawn } = require( "child_process" );
const { argv, nextTick, stderr, stdout } = require( "process" );

( async () => {
    if ( argv.length < 3 ) { return; }
    await new Promise( ( resolve, reject ) => {
        console.log( argv.slice( 2 ) );
        const child = spawn( argv[ 2 ], argv.slice( 3 ), {
            detached: true
        });
        child.unref();
        child.on( "error", reject );
        child.stderr.pipe( stderr );
        child.stdout.pipe( stdout );
        nextTick( resolve );
    });
    await new Promise( resolve => setTimeout( resolve, 4000 ) );
})();
```

**foo.ps1**
```pwsh
for($i=0; $i -lt $args.length; $i++) {
  "Arg " + $i + ": " + $args[$i]
}

explorer
```

If you run `foo.ps1` like this, it'll print its argv to stdout and launch explorer.exe:
```sh
$ powershell -ExecutionPolicy ByPass -File .\foo.ps1 foo bar
Arg 0: foo
Arg 1: bar
```

But if you run it from NodeJS with `detached: true`, it won't work:
```sh
$ node .\foo.js powershell -ExecutionPolicy ByPass -File .\foo.ps1 foo bar
[
  'powershell',
  '-ExecutionPolicy',
  'ByPass',
  '-File',
  '.\\foo.ps1',
  'foo',
  'bar'
]
```

Setting `detached: false` (or uncommenting the line) however works:
```sh
$ node .\foo.js powershell -ExecutionPolicy ByPass -File .\foo.ps1 foo bar
[
  'powershell',
  '-ExecutionPolicy',
  'ByPass',
  '-File',
  '.\\foo.ps1',
  'foo',
  'bar'
]
Arg 0: foo
Arg 1: bar
```

Using the `cmd.exe` wrapper also works from NodeJS (stdio will be ignored though):
```sh
$ node .\foo.js cmd /c 'powershell -ExecutionPolicy ByPass -File .\foo.ps1 foo bar'
[
  'cmd',
  '/c',
  'powershell -ExecutionPolicy ByPass -File .\\foo.ps1 foo bar'
]
```
